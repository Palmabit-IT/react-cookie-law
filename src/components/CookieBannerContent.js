import React from 'react';
import bannerStyle from './bannerStyle';
import CookieBannerPreferences from './CookieBannerPreferences';

class CookieBannerContent extends React.Component {
  constructor(props) {
    super(props);

    this.state = { showPreferences: false };
  }

  render() {
    const {
      styles = {},
      className = '',
      message = 'No text',
      policyLink = '/#',
      privacyPolicyLinkText = 'Privacy Policy',
      acceptButtonText = 'Accept all',
      managePreferencesButtonText = 'Mange my cookies',
      savePreferencesButtonText = 'Save and close',
      onConfirm = () => {},
      onAcceptAll = () => {},
    } = this.props;

    const { showPreferences } = this.state;

    const {
      dialog: dialogStyle,
      container: containerStyle,
      message: messageStyle,
      policy: policyStyle,
      buttonWrapper: buttonWrapperStyle,
      button: buttonStyle,
    } = { ...bannerStyle, ...styles };

    return (
      <div
        className={`react-cookie-law-dialog ${className}`}
        style={dialogStyle}
      >
        <div className="react-cookie-law-container" style={containerStyle}>
          <div className="react-cookie-law-msg" style={messageStyle}>
            {message}
          </div>

          {showPreferences && <CookieBannerPreferences {...this.props} />}

          <a
            href={policyLink}
            target="_blank"
            className="react-cookie-law-policy"
            style={policyStyle}
          >
            {privacyPolicyLinkText}
          </a>

          <div
            className="react-cookie-law-button-wrapper"
            style={buttonWrapperStyle}
          >
            {showPreferences ? (
              <button
                type="button"
                className="react-cookie-law-save-btn"
                style={buttonStyle}
                onClick={() => onConfirm()}
              >
                <span>{savePreferencesButtonText}</span>
              </button>
            ) : (
              <button
                type="button"
                className="react-cookie-law-manage-btn"
                style={buttonStyle}
                onClick={() => this.setState({ showPreferences: true })}
              >
                <span>{managePreferencesButtonText}</span>
              </button>
            )}
            <button
              type="button"
              className="react-cookie-law-accept-btn"
              style={buttonStyle}
              onClick={() => onAcceptAll()}
            >
              <span>{acceptButtonText}</span>
            </button>
          </div>
        </div>
      </div>
    );
  }
}

export default CookieBannerContent;
