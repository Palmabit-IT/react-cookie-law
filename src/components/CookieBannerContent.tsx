import React from 'react';
import bannerStyle from './bannerStyle';
import CookieBannerPreferences from './CookieBannerPreferences';

interface Props extends React.ComponentProps<typeof CookieBannerPreferences> {
  acceptButtonText?: string
  className?: string
  // Not used to-date.
  declineButtonText?: string
  managePreferencesButtonText?: string
  message?: string
  onAcceptAll?: () => void
  // Not used to-date.
  onDecline?: () => void
  onConfirm?: () => void
  policyLink?: string
  privacyPolicyLinkText?: string
  savePreferencesButtonText?: string
  // Not used to-date.
  showDeclineButton?: boolean
}

interface State {
  showPreferences?: boolean
}
class CookieBannerContent extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = { showPreferences: false };
  }

  render() {
    const {
      acceptButtonText = 'Accept all',
      className = '',
      managePreferencesButtonText = 'Mange my cookies',
      message = 'No text',
      onAcceptAll = () => undefined,
      onConfirm = () => undefined,
      policyLink = '/#',
      privacyPolicyLinkText = 'Privacy Policy',
      savePreferencesButtonText = 'Save and close',
      ...preferenceProps
    } = this.props;

    const { showPreferences } = this.state;

    const {
      dialog: dialogStyle,
      container: containerStyle,
      message: messageStyle,
      policy: policyStyle,
      buttonWrapper: buttonWrapperStyle,
      button: buttonStyle,
    } = { ...bannerStyle, ...preferenceProps.styles };

    return (
      <div
        className={`react-cookie-law-dialog ${className}`}
        style={dialogStyle}
      >
        <div className="react-cookie-law-container" style={containerStyle}>
          <div className="react-cookie-law-msg" style={messageStyle}>
            {message}
          </div>

          {showPreferences && <CookieBannerPreferences {...preferenceProps} />}

          <a
            href={policyLink}
            className="react-cookie-law-policy"
            style={policyStyle}
          >
            {privacyPolicyLinkText}
          </a>

          <div
            className="react-cookie-law-button-wrapper"
            style={buttonWrapperStyle}
          >
            {showPreferences ? (
              <button
                type="button"
                className="react-cookie-law-save-btn"
                style={buttonStyle}
                onClick={() => onConfirm()}
              >
                <span>{savePreferencesButtonText}</span>
              </button>
            ) : (
              <button
                type="button"
                className="react-cookie-law-manage-btn"
                style={buttonStyle}
                onClick={() => this.setState({ showPreferences: true })}
              >
                <span>{managePreferencesButtonText}</span>
              </button>
            )}
            <button
              type="button"
              className="react-cookie-law-accept-btn"
              style={buttonStyle}
              onClick={() => onAcceptAll()}
            >
              <span>{acceptButtonText}</span>
            </button>
          </div>
        </div>
      </div>
    );
  }
}

export default CookieBannerContent;
