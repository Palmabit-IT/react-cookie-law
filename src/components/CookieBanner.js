import React from 'react';
import PropTypes from 'prop-types';
import Cookies from '../Cookies';
import CookieBannerContent from './CookieBannerContent';
import { isServer } from '../helpers';

const CONSENT_GIVEN = 'rcl_consent_given';
const PREFERENCES_COOKIE = 'rcl_preferences_consent';
const STATISTICS_COOKIE = 'rcl_statistics_consent';
const MARKETING_COOKIE = 'rcl_marketing_consent';

class CookieBanner extends React.Component {
  constructor(props) {
    super(props);

    const {
      preferencesDefaultChecked = false,
      statisticsDefaultChecked = false,
      marketingDefaultChecked = false,
      wholeDomain = false,
    } = this.props;

    this.state = {
      preferencesCookies: preferencesDefaultChecked,
      statisticsCookies: statisticsDefaultChecked,
      marketingCookies: marketingDefaultChecked,
    };

    this.onScroll = this.onScroll.bind(this);
    this.onTogglePreferencesCookies = this.onTogglePreferencesCookies.bind(this);
    this.onToggleStatisticsCookies = this.onToggleStatisticsCookies.bind(this);
    this.onToggleMarketingCookies = this.onToggleMarketingCookies.bind(this);
    this.onAcceptAll = this.onAcceptAll.bind(this);
    this.confirm = this.confirm.bind(this);
    this.decline = this.decline.bind(this);
    this.consetsCallback = this.consetsCallback.bind(this);

    this.cookies = new Cookies(wholeDomain);
  }

  componentDidMount() {
    const { dismissOnScroll } = this.props;

    if (isServer() || dismissOnScroll !== true) {
      return;
    }

    if (window.addEventListener) {
      window.addEventListener('scroll', this.onScroll);
    } else if (window.attachEvent) {
      window.attachEvent('onscroll', this.onScroll); // < IE9
    }
  }

  componentWillUnmount() {
    if (isServer()) {
      return;
    }

    if (window.removeEventListener) {
      window.removeEventListener('scroll', this.onScroll);
    } else if (window.detachEvent) {
      window.detachEvent('onscroll', this.onScroll); // < IE9
    }
  }

  onScroll() {
    this.confirm();
  }

  onTogglePreferencesCookies(value) {
    this.setState({ preferencesCookies: value });
  }

  onToggleStatisticsCookies(value) {
    this.setState({ statisticsCookies: value });
  }

  onToggleMarketingCookies(value) {
    this.setState({ marketingCookies: value });
  }

  onAcceptAll() {
    this.cookies.set(CONSENT_GIVEN);
    this.cookies.set(PREFERENCES_COOKIE);
    this.cookies.set(STATISTICS_COOKIE);
    this.cookies.set(MARKETING_COOKIE);

    this.consetsCallback();

    this.forceUpdate();
  }

  confirm() {
    const { preferencesCookies, statisticsCookies, marketingCookies } = this.state;

    this.cookies.set(CONSENT_GIVEN);

    if (preferencesCookies) {
      this.cookies.set(PREFERENCES_COOKIE);
    } else {
      this.cookies.remove(PREFERENCES_COOKIE);
    }

    if (statisticsCookies) {
      this.cookies.set(STATISTICS_COOKIE);
    } else {
      this.cookies.remove(STATISTICS_COOKIE);
    }

    if (marketingCookies) {
      this.cookies.set(MARKETING_COOKIE);
    } else {
      this.cookies.remove(MARKETING_COOKIE);
    }

    this.consetsCallback();

    this.forceUpdate();
  }

  decline() {
    const {
      onDeclinePreferences = () => {},
      onDeclineStatistics = () => {},
      onDeclineMarketing = () => {},
    } = this.props;

    this.cookies.set(CONSENT_GIVEN);
    this.cookies.remove(PREFERENCES_COOKIE);
    this.cookies.remove(STATISTICS_COOKIE);
    this.cookies.remove(MARKETING_COOKIE);

    onDeclinePreferences();
    onDeclineStatistics();
    onDeclineMarketing();

    this.forceUpdate();
  }

  consetsCallback() {
    const {
      onAccept = () => {},
      onAcceptPreferences = () => {},
      onAcceptStatistics = () => {},
      onAcceptMarketing = () => {},
      onDeclinePreferences = () => {},
      onDeclineStatistics = () => {},
      onDeclineMarketing = () => {},
    } = this.props;

    const hasPreferencesCookie = this.cookies.get(PREFERENCES_COOKIE);
    const hasStatisticsCookie = this.cookies.get(STATISTICS_COOKIE);
    const hasMarketingCookie = this.cookies.get(MARKETING_COOKIE);

    onAccept();

    if (hasPreferencesCookie) {
      onAcceptPreferences();
    } else {
      onDeclinePreferences();
    }

    if (hasStatisticsCookie) {
      onAcceptStatistics();
    } else {
      onDeclineStatistics();
    }

    if (hasMarketingCookie) {
      onAcceptMarketing();
    } else {
      onDeclineMarketing();
    }
  }

  render() {
    const {
      styles,
      className,
      message,
      policyLink,
      privacyPolicyLinkText,
      necessaryOptionText,
      preferencesOptionText,
      statisticsOptionText,
      marketingOptionText,
      showDeclineButton,
      acceptButtonText,
      declineButtonText,
      managePreferencesButtonText,
      savePreferencesButtonText,
      showPreferencesOption,
      showStatisticsOption,
      showMarketingOption,
      preferencesDefaultChecked,
      statisticsDefaultChecked,
      marketingDefaultChecked,
    } = this.props;

    if (this.cookies.get(CONSENT_GIVEN)) {
      return null;
    }

    const contentProps = {
      styles,
      className,
      message,
      policyLink,
      privacyPolicyLinkText,
      necessaryOptionText,
      preferencesOptionText,
      statisticsOptionText,
      marketingOptionText,
      showDeclineButton,
      acceptButtonText,
      declineButtonText,
      managePreferencesButtonText,
      savePreferencesButtonText,
      showPreferencesOption,
      showStatisticsOption,
      showMarketingOption,
      preferencesDefaultChecked,
      statisticsDefaultChecked,
      marketingDefaultChecked,
      onTogglePreferencesCookies: this.onTogglePreferencesCookies,
      onToggleStatisticsCookies: this.onToggleStatisticsCookies,
      onToggleMarketingCookies: this.onToggleMarketingCookies,
      onDecline: this.decline,
      onConfirm: this.confirm,
      onAcceptAll: this.onAcceptAll,
    };

    return (<CookieBannerContent {...contentProps} />);
  }
}

CookieBanner.protoTypes = {
  className: PropTypes.string,
  styles: PropTypes.object,
  message: PropTypes.string.isRequired,
  wholeDomain: PropTypes.bool,
  policyLink: PropTypes.string,
  privacyPolicyLinkText: PropTypes.string,
  necessaryOptionText: PropTypes.string,
  preferencesOptionText: PropTypes.string,
  statisticsOptionText: PropTypes.string,
  marketingOptionText: PropTypes.string,
  acceptButtonText: PropTypes.string,
  declineButtonText: PropTypes.string,
  managePreferencesButtonText: PropTypes.string,
  savePreferencesButtonText: PropTypes.string,
  showDeclineButton: PropTypes.bool,
  dismissOnScroll: PropTypes.bool,
  showPreferencesOption: PropTypes.bool,
  showStatisticsOption: PropTypes.bool,
  showMarketingOption: PropTypes.bool,
  preferencesDefaultChecked: PropTypes.bool,
  statisticsDefaultChecked: PropTypes.bool,
  marketingDefaultChecked: PropTypes.bool,
  onAccept: PropTypes.func,
  onAcceptPreferences: PropTypes.func,
  onAcceptStatistics: PropTypes.func,
  onAcceptMarketing: PropTypes.func,
  onDeclinePreferences: PropTypes.func,
  onDeclineStatistics: PropTypes.func,
  onDeclineMarketing: PropTypes.func,
};

export default CookieBanner;
