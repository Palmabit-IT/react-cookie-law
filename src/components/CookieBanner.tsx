import React, { ComponentProps } from 'react';
import slugify from 'slugify';
import { v4 as uuidv4 } from 'uuid';

import Cookies from '../Cookies';
import CookieBannerContent from './CookieBannerContent';
import { CoryphaApiProps, isServer, isUsingCorypha } from '../helpers';
import {
  CoryphaPreference,
  fetchCoryphaPreferences,
  saveCoryphaPreferences,
  checkVersionCoryphaPreferences,
} from '../coryphaAPI';

const CORYPHA_USER = 'rl_corypha_user';
const CONSENT_GIVEN = 'rcl_consent_given';
const PREFERENCES_COOKIE = 'rcl_preferences_consent';
const STATISTICS_COOKIE = 'rcl_statistics_consent';
const MARKETING_COOKIE = 'rcl_marketing_consent';

interface Props extends CoryphaApiProps, ComponentProps<typeof CookieBannerContent> {
  preferencesDefaultChecked?: boolean
  statisticsDefaultChecked?: boolean
  marketingDefaultChecked?: boolean
  dismissOnScroll?: boolean
  onAccept?: () => void
  onAcceptCoryphaPreferences?: (preferences: CoryphaPreference[]) => void
  onAcceptMarketing?: () => void
  onAcceptPreferences?: () => void
  onAcceptStatistics?: () => void
  onDeclineMarketing?: () => void
  onDeclinePreferences?: () => void
  onDeclineStatistics?: () => void
  onDeclineCoryphaPreferences?: (preferences: CoryphaPreference[]) => void
}

interface State {
  preferencesCookies: boolean
  statisticsCookies: boolean
  marketingCookies: boolean
  coryphaPreferences: CoryphaPreference[]
}
class CookieBanner extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    const {
      preferencesDefaultChecked = false,
      statisticsDefaultChecked = false,
      marketingDefaultChecked = false,
      wholeDomain = false,
    } = this.props;

    this.state = {
      preferencesCookies: preferencesDefaultChecked,
      statisticsCookies: statisticsDefaultChecked,
      marketingCookies: marketingDefaultChecked,
      coryphaPreferences: [],
    };

    this.onScroll = this.onScroll.bind(this);
    this.onTogglePreferencesCookies = this.onTogglePreferencesCookies.bind(this);
    this.onToggleStatisticsCookies = this.onToggleStatisticsCookies.bind(this);
    this.onToggleMarketingCookies = this.onToggleMarketingCookies.bind(this);
    this.onAcceptAll = this.onAcceptAll.bind(this);
    this.confirm = this.confirm.bind(this);
    this.decline = this.decline.bind(this);
    this.consetsCallback = this.consetsCallback.bind(this);
    this.onToggleCoryphaPreference = this.onToggleCoryphaPreference.bind(this);

    this.cookies = new Cookies(wholeDomain);
  }

  async componentDidMount() {
    const { dismissOnScroll } = this.props;

    if (isUsingCorypha(this.props)) {
      if (this.cookies.get(CONSENT_GIVEN)) {
        const { hasNewVersion, preferences } = await checkVersionCoryphaPreferences(this.props);

        if (hasNewVersion) {
          this.cookies.clear();

          this.setState({ coryphaPreferences: preferences }, () => {
            this.forceUpdate();
          });
        }
      } else {
        const preferences = await fetchCoryphaPreferences(this.props);

        this.setState({
          coryphaPreferences: preferences.map((preference) => ({
            ...preference,
            accepted: preference.required,
          })),
        });
      }
    }

    if (isServer() || dismissOnScroll !== true) {
      return;
    }

    if (window.addEventListener) {
      window.addEventListener('scroll', this.onScroll);
    } else if (window.attachEvent) {
      window.attachEvent('onscroll', this.onScroll); // < IE9
    }
  }

  componentWillUnmount() {
    if (isServer()) {
      return;
    }

    if (window.removeEventListener) {
      window.removeEventListener('scroll', this.onScroll);
    } else if (window.detachEvent) {
      window.detachEvent('onscroll', this.onScroll); // < IE9
    }
  }

  onScroll() {
    this.confirm();
  }

  onTogglePreferencesCookies(value?: boolean) {
    this.setState({ preferencesCookies: value });
  }

  onToggleStatisticsCookies(value?: boolean) {
    this.setState({ statisticsCookies: value });
  }

  onToggleMarketingCookies(value?: boolean) {
    this.setState({ marketingCookies: value });
  }

  onToggleCoryphaPreference(value, preference) {
    const { coryphaPreferences = [] } = this.state;
    const newCoryphaPreferences = [...coryphaPreferences];
    const changedPreference = newCoryphaPreferences.find(
      ({ id }) => id === preference.id,
    );

    changedPreference.accepted = value;

    this.setState({ coryphaPreferences: newCoryphaPreferences });
  }

  onAcceptAll() {
    const {
      onAcceptPreferences = () => undefined,
      onAcceptStatistics = () => undefined,
      onAcceptMarketing = () => undefined,
      onAcceptCoryphaPreferences = () => undefined,
      coryphaUserId,
    } = this.props;

    const { coryphaPreferences = [] } = this.state;

    if (isUsingCorypha(this.props)) {
      this.cookies.set(CONSENT_GIVEN);
      this.cookies.set({
        key: CORYPHA_USER,
        value: coryphaUserId || uuidv4(),
      });

      coryphaPreferences.forEach(({ name }) => {
        this.cookies.set(`rl_${slugify(name.toLowerCase(), '_')}_consent`);
      });

      onAcceptCoryphaPreferences(coryphaPreferences);
      saveCoryphaPreferences(
        this.props,
        coryphaPreferences.map((preference) => ({
          id: preference.id,
          accepted: true,
        })),
      );
    } else {
      this.cookies.set(CONSENT_GIVEN);
      this.cookies.set(PREFERENCES_COOKIE);
      this.cookies.set(STATISTICS_COOKIE);
      this.cookies.set(MARKETING_COOKIE);

      onAcceptPreferences();
      onAcceptStatistics();
      onAcceptMarketing();
    }

    this.forceUpdate();
  }

  cookies: Cookies

  confirm() {
    const { coryphaUserId } = this.props;
    const { coryphaPreferences = [] } = this.state;

    if (isUsingCorypha(this.props)) {
      this.cookies.set(CONSENT_GIVEN);
      this.cookies.set({
        key: CORYPHA_USER,
        value: coryphaUserId || uuidv4(),
      });

      coryphaPreferences.forEach(({ name, accepted }) => {
        this.cookies.set({
          key: `rl_${slugify(name.toLowerCase(), '_')}_consent`,
          value: accepted,
        });
      });

      saveCoryphaPreferences(
        this.props,
        coryphaPreferences.map((preference) => ({
          id: preference.id,
          accepted: preference.accepted,
        })),
      );
    } else {
      const { preferencesCookies, statisticsCookies, marketingCookies } = this.state;

      this.cookies.set(CONSENT_GIVEN);

      if (preferencesCookies) {
        this.cookies.set(PREFERENCES_COOKIE);
      } else {
        this.cookies.remove(PREFERENCES_COOKIE);
      }

      if (statisticsCookies) {
        this.cookies.set(STATISTICS_COOKIE);
      } else {
        this.cookies.remove(STATISTICS_COOKIE);
      }

      if (marketingCookies) {
        this.cookies.set(MARKETING_COOKIE);
      } else {
        this.cookies.remove(MARKETING_COOKIE);
      }
    }

    this.forceUpdate();
  }

  decline() {
    const {
      onDeclinePreferences = () => undefined,
      onDeclineStatistics = () => undefined,
      onDeclineMarketing = () => undefined,
      onDeclineCoryphaPreferences = () => undefined,
    } = this.props;

    const { coryphaPreferences = [] } = this.state;

    if (isUsingCorypha(this.props)) {
      this.cookies.set(CONSENT_GIVEN);
      this.cookies.remove(CORYPHA_USER);

      coryphaPreferences.forEach(({ name }) => {
        this.cookies.remove(`rl_${slugify(name.toLowerCase(), '_')}_consent`);
      });

      onDeclineCoryphaPreferences(coryphaPreferences);
    } else {
      this.cookies.set(CONSENT_GIVEN);
      this.cookies.remove(PREFERENCES_COOKIE);
      this.cookies.remove(STATISTICS_COOKIE);
      this.cookies.remove(MARKETING_COOKIE);

      onDeclinePreferences();
      onDeclineStatistics();
      onDeclineMarketing();
    }

    this.forceUpdate();
  }

  consetsCallback() {
    const {
      onAccept = () => undefined,
      onAcceptPreferences = () => undefined,
      onAcceptStatistics = () => undefined,
      onAcceptMarketing = () => undefined,
      onDeclinePreferences = () => undefined,
      onDeclineStatistics = () => undefined,
      onDeclineMarketing = () => undefined,
    } = this.props;

    const hasPreferencesCookie = this.cookies.get(PREFERENCES_COOKIE);
    const hasStatisticsCookie = this.cookies.get(STATISTICS_COOKIE);
    const hasMarketingCookie = this.cookies.get(MARKETING_COOKIE);

    onAccept();

    if (hasPreferencesCookie) {
      onAcceptPreferences();
    } else {
      onDeclinePreferences();
    }

    if (hasStatisticsCookie) {
      onAcceptStatistics();
    } else {
      onDeclineStatistics();
    }

    if (hasMarketingCookie) {
      onAcceptMarketing();
    } else {
      onDeclineMarketing();
    }
  }

  render() {
    const {
      styles,
      className,
      message,
      policyLink,
      privacyPolicyLinkText,
      necessaryOptionText,
      preferencesOptionText,
      statisticsOptionText,
      marketingOptionText,
      showDeclineButton,
      acceptButtonText,
      declineButtonText,
      managePreferencesButtonText,
      savePreferencesButtonText,
      showPreferencesOption,
      showStatisticsOption,
      showMarketingOption,
      preferencesDefaultChecked,
      statisticsDefaultChecked,
      marketingDefaultChecked,
    } = this.props;

    const { coryphaPreferences = [] } = this.state;

    if (this.cookies.get(CONSENT_GIVEN)) {
      this.consetsCallback();
      return null;
    }

    const contentProps = {
      styles,
      className,
      message,
      policyLink,
      privacyPolicyLinkText,
      necessaryOptionText,
      preferencesOptionText,
      statisticsOptionText,
      marketingOptionText,
      showDeclineButton,
      acceptButtonText,
      declineButtonText,
      managePreferencesButtonText,
      savePreferencesButtonText,
      showPreferencesOption,
      showStatisticsOption,
      showMarketingOption,
      preferencesDefaultChecked,
      statisticsDefaultChecked,
      marketingDefaultChecked,
      onTogglePreferencesCookies: this.onTogglePreferencesCookies,
      onToggleStatisticsCookies: this.onToggleStatisticsCookies,
      onToggleMarketingCookies: this.onToggleMarketingCookies,
      onDecline: this.decline,
      onConfirm: this.confirm,
      onAcceptAll: this.onAcceptAll,
      coryphaPreferences,
      onToggleCoryphaPreference: this.onToggleCoryphaPreference,
    };

    return <CookieBannerContent {...contentProps} />;
  }
}

export default CookieBanner;
