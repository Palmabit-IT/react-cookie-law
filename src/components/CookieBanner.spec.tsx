import React from 'react';
import { mount, shallow } from 'enzyme';

import CookieBanner from './CookieBanner';
import CookieBannerContent from './CookieBannerContent';

describe('CookieBanner component', () => {
  beforeEach(() => {
    document.cookie.split(';').forEach((cookie) => {
      const eqPos = cookie.indexOf('=');
      const name = eqPos > -1 ? cookie.substr(0, eqPos) : cookie;
      document.cookie = `${name}=;expires=Thu, 01 Jan 1970 00:00:00 GMT`;
    });
  });

  test('should be rendered', () => {
    const component = shallow(
      <CookieBanner message="Text" />,
    );

    expect(component).toMatchSnapshot();
  });

  test('should be rendered with elements', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    expect(component.find(CookieBannerContent)).toHaveLength(1);
  });

  test('should be rendered with custom class', () => {
    const component = mount(
      <CookieBanner className="testclass" message="Custom text" />,
    );

    expect(component.hasClass('testclass')).toBe(true);
  });

  test('shouldn\'t show banner if cookies are already accepted', () => {
    document.cookie = 'rcl_consent_given=true';

    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    expect(component.find(CookieBannerContent)).toHaveLength(0);
  });

  test('should close banner if confirm function is called', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    expect(component.html().length).toBeGreaterThan(0);
    expect(component.find(CookieBannerContent)).toHaveLength(1);
    component.find(CookieBannerContent).prop('onConfirm')();
  });

  test('should close banner if decline button is pressed', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    expect(component.html().length).toBeGreaterThan(0);
    expect(component.find(CookieBannerContent)).toHaveLength(1);
    component.find(CookieBannerContent).prop('onDecline')();
    // expect(component.html()).toBeNull();
  });

  test('should close banner at scrolling if dismissOnScroll is true', () => {
    const component = mount(
      <CookieBanner message="Test" dismissOnScroll />,
    );

    expect(component.html().length).toBeGreaterThan(0);
    window.dispatchEvent(new window.UIEvent('scroll', { detail: 0 }));
    // expect(component.html()).toBeNull();
  });

  test('shouldn\'t close banner at scrolling if dismissOnScroll is true', () => {
    const component = mount(
      <CookieBanner message="Test" />,
    );

    expect(component.html().length).toBeGreaterThan(0);
    window.dispatchEvent(new window.UIEvent('scroll', { detail: 0 }));
    expect(component.html().length).toBeGreaterThan(0);
  });

  test('should give consent if confirm function is called', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onConfirm')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
  });

  test('should give consent if decline function is called', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onDecline')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
  });

  test('should give consent at scrolling if dismissOnScroll is true', () => {
    mount(<CookieBanner message="Custom text" dismissOnScroll />);

    window.dispatchEvent(new window.UIEvent('scroll', { detail: 0 }));
    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
  });

  test('should accept all cookies', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onAcceptAll')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_preferences_consent=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_statistics_consent=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_marketing_consent=true')).toBeGreaterThanOrEqual(0);
  });

  test('should give default consents at confirm', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onConfirm')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_preferences_consent=true')).toBeLessThan(0);
    expect(document.cookie.indexOf('rcl_statistics_consent=true')).toBeLessThan(0);
    expect(document.cookie.indexOf('rcl_marketing_consent=true')).toBeLessThan(0);
  });

  test('should remove consent for preferences cookie at confirm', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onTogglePreferencesCookies')(true);
    component.find(CookieBannerContent).prop('onConfirm')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_preferences_consent=true')).toBeGreaterThanOrEqual(0);
  });

  test('should give consent for preferences cookie at confirm', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onTogglePreferencesCookies')(true);
    component.find(CookieBannerContent).prop('onTogglePreferencesCookies')(false);
    component.find(CookieBannerContent).prop('onConfirm')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_preferences_consent=true')).toBeLessThan(0);
  });

  test('should remove consent for statistics cookie at confirm', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onToggleStatisticsCookies')(true);
    component.find(CookieBannerContent).prop('onConfirm')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_statistics_consent=true')).toBeGreaterThanOrEqual(0);
  });

  test('should give consent for statistics cookie at confirm', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onToggleStatisticsCookies')(true);
    component.find(CookieBannerContent).prop('onToggleStatisticsCookies')(false);
    component.find(CookieBannerContent).prop('onConfirm')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_statistics_consent=true')).toBeLessThan(0);
  });

  test('should give consent for marketing cookie at confirm', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onToggleMarketingCookies')(true);
    component.find(CookieBannerContent).prop('onConfirm')();

    expect(document.cookie.indexOf('rcl_consent_given=true')).toBeGreaterThanOrEqual(0);
    expect(document.cookie.indexOf('rcl_marketing_consent=true')).toBeGreaterThanOrEqual(0);
  });

  test('shouldn\'t give consent at decline', () => {
    const component = mount(
      <CookieBanner message="Custom text" />,
    );

    component.find(CookieBannerContent).prop('onTogglePreferencesCookies')(true);
    component.find(CookieBannerContent).prop('onToggleStatisticsCookies')(true);
    component.find(CookieBannerContent).prop('onToggleMarketingCookies')(true);
    component.find(CookieBannerContent).prop('onDecline')();

    expect(document.cookie).toBe('rcl_consent_given=true');
    expect(document.cookie.indexOf('rcl_preferences_consent=true')).toBeLessThan(0);
    expect(document.cookie.indexOf('rcl_statistics_consent=true')).toBeLessThan(0);
    expect(document.cookie.indexOf('rcl_marketing_consent=true')).toBeLessThan(0);
  });
});
