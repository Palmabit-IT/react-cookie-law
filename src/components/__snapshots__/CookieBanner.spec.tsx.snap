// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieBanner component should be rendered 1`] = `
<CookieBannerContent
  coryphaPreferences={Array []}
  message="Text"
  onAcceptAll={[Function]}
  onConfirm={[Function]}
  onDecline={[Function]}
  onToggleCoryphaPreference={[Function]}
  onToggleMarketingCookies={[Function]}
  onTogglePreferencesCookies={[Function]}
  onToggleStatisticsCookies={[Function]}
/>
`;
